enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,155 @@
+
+## Summary
+
+The purpose of this enhancement is to outline how the [Operator Lifecycle Manager (OLM)][olm] can let operator users opt-in to the clean up of their operands, or operator workloads, upon the uninstallation of their operator. Specifically when uninstalling an operator, OLM would first delete all CustomResources (CRs) managed by an operator to provide the operator time to clean up all associated resources before deleting the operator itself.
+
+## Motivation
+
+By design when OLM uninstalls an operator it does not remove any of the operator’s owned CustomResourceDefinitions (CRDs) and CRs in order to prevent data loss.
+The operator user or cluster admin is expected to manually delete CRDs, CRs and any other secondary resources created by the operator.
+While some secondary resources that have their [OwnerReference][owner-reference] set to a CR would be garbage collected, others such as off-cluster or cross-namespaced resources with no owner references would remain orphaned.
+
+Without the operator around to process deletion events of CRs, it becomes hard to cleanly remove deployed resources, especially when they exist outside of the cluster.
+
+### Goals
+
+- Clean up includes removal of CRs (operands) and CRDs
+- Operands and CRDs are not automatically removed
+- Upon operator uninstall a user can opt-in to the clean up behavior
+- The operand removal should have a timeout
+- After the operand removal timeout the operator should still be uninstalled
+- Exceeding the operand removal timeout should create an alert event
+
+### Non-Goals
+
+- OLM handles cleanup of anything other than CRs and CRDs
+- Handle custom cleanup logic like filtering/ordering the deletion of CRs 
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator user I can instruct OLM to clean up all of my operator's CRs and CRDs when I uninstall my operator.
+
+#### Story 2
+
+As an operator user I want to see an alert event when OLM does not succeed in operand cleanup after a configured timeout.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Specify cleanup via Subscription
+
+To provide a way to opt-in for operand cleanup, the [`Subscription`][subscription] spec can include an optional field `spec.operandCleanup` which lets a user configure the cleanup behaviour.
+
+OLM would initiate cleanup upon the deletion of a `Subscription` instance, based on `spec.operandCleanup`:
+
+##### None (default)
+
+With `spec.operandCleanup.type: none` OLM will not clean up operands. This is also the default behavior if `spec.operandCleanup` is unspecified to make clean up opt-in.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+spec:
+  ...
+  operandCleanup:
+    type: none
+```
+
+##### OLM managed
+
+With `operandCleanup.type: olm-managed` OLM will delete all CRs, for the operator's owned APIs, in the target namespaces (`olm.targetNamespaces`) of that operator.
```

Comment:
Can you elaborate on why OLM should delete all CRs before the CRD? Deleting a CRD will garbage collect all CRs. An operator could just as easily place a finalizer on its CRD to prevent this until it's finished doing whatever other work it needs to do.

Reply:
> Deleting a CRD will garbage collect all CRs

This would become an issue when you have two or more instances of a non-cluster scoped operator installed (targeting different own/single/multi namespaces). 
If subscription A opts-in to cleanup while subscription B doesn't, then upon uninstalling A you'd remove the CRD, which would also garbage collect all the CRs being managed by operator B, while it's still running.

I'll update the proposal to clarify this point though. 


> An operator could just as easily place a finalizer on its CRD

Not sure if you mean CRD as in the `kind: CustomResourceDefinition` object, or the custom resource instance (`kind: FooBar`) which I'm referring to as CR here.

If the former then I don't think operators (especially non-cluster scoped ones) usually include permissions for setting finalizers on the `CustomResourceDefinition` object.
If the latter, then I agree. That's the expectation here: OLM deletes the CR instance, but that blocks because the operator has a finalizer on it.

Reply:
Forgot I already have a [section below](https://github.com/operator-framework/enhancements/pull/46/files#diff-3824b2623f1ada24ff884541902a3edaR117-R122) on `cleanup for namespace-scoped operators` which outlines when and why OLM might not delete the CRD as part of cleanup. Let me know if that part still needs clarification.

Reply:
> This would become an issue when you have two or more instances of a non-cluster scoped operator installed (targeting different own/single/multi namespaces).

Good point! Also this is an issue with any non-global operator, installed more than once, which has _cluster-scoped CRs_ since we allow this today.

> Not sure if you mean CRD as in the kind: CustomResourceDefinition object,

I mean CRD, not CR

> If the former then I don't think operators (especially non-cluster scoped ones) usually include permissions for setting finalizers on the CustomResourceDefinition object.

Trying to make a similar point to https://github.com/operator-framework/enhancements/pull/46/files#r483578282 -- your first comment about scoping pokes a hole in it (for so long as non-global operators are allowed).